useful-scripts/permalink-snapshot-migration/migrate_legacy_permalinks.py

```diff
@@ -0,0 +1,114 @@
+import argparse
+import csv
+import os
+import time
+
+import requests
+
+"""
+This is the script for migrating legacy permalinks to permalink snapshots.
+It uses a Data API migration endpoint `PUT /api/permalink/{permalinkId}/snapshot`.
+It reads a CSV file named 'permalinks.csv' and outputs a CSV file named 'migration_results.csv'.
+
+Usage: `pipenv run python migrate_legacy_permalinks.py`
+
+Instructions:
+
+1. List all the legacy permalinks by running a query against the Content database to get all the legacy permalinks:
+
+```
+SELECT LOWER(Id) AS permalink_id,
+       IIF(LegacyHasSnapshot = 1, 'True', 'False') AS migrated
+FROM content.dbo.Permalinks
+WHERE Legacy = 1
+ORDER BY Created
```

Comment:
Minor point that frankly can probably be ignored, but this will mean we migrate the oldest permalink first? Not sure how long the entire migration will take, but if it'll take a while we could `order by desc` so that we migrate those permalinks most likely to be used first.

Reply:
The front-end will have a feature toggle that won't be removed until we've migrated all permalinks and verified that migration by comparing before/after results using both routes.

I don't think there's a need to make sure we migrate those likely to be used first as we are trying to achieve 100% through identifying ones we think are problematic early through other tickets and addressing their issues or deleting them.

When we come to actually run it we might not care about any migration order, or might want to migrate them in size order instead.

The idea is that they will all be migrated by the time the old routes are removed.